types/wlr_keyboard.c

```diff
@@ -21,37 +21,12 @@ static void keyboard_led_update(struct wlr_keyboard *keyboard) {
 	wlr_keyboard_led_update(keyboard, leds);
 }
 
-static void keyboard_modifier_update(struct wlr_keyboard *keyboard) {
-	xkb_mod_mask_t depressed = xkb_state_serialize_mods(keyboard->xkb_state,
-		XKB_STATE_DEPRESSED);
-	xkb_mod_mask_t latched = xkb_state_serialize_mods(keyboard->xkb_state,
-		XKB_STATE_LATCHED);
-	xkb_mod_mask_t locked = xkb_state_serialize_mods(keyboard->xkb_state,
-		XKB_STATE_LOCKED);
-	xkb_mod_mask_t group = xkb_state_serialize_layout(keyboard->xkb_state,
-		XKB_STATE_LAYOUT_EFFECTIVE);
-	if (depressed == keyboard->modifiers.depressed &&
-			latched == keyboard->modifiers.latched &&
-			locked == keyboard->modifiers.locked &&
-			group == keyboard->modifiers.group) {
-		return;
-	}
-
-	keyboard->modifiers.depressed = depressed;
-	keyboard->modifiers.latched = latched;
-	keyboard->modifiers.locked = locked;
-	keyboard->modifiers.group = group;
-
-	// TODO: wl_keyboard_send_modifiers
```

Comment:
I was just referring to wl_keyboard_send_modifiers when I said you don't have to do this. You still have to update depressed/latched/locked/group, or else wlr_keyboard_get_modifiers doesn't work without a seat. You should update seat to not duplicate the work, rather than pulling it out of here. Maybe raise a modifiers changed signal?

Reply:
Yeah, this is now done in the seat: https://github.com/swaywm/wlroots/pull/184/files#diff-259a69ede9a79f9859eda604dd141090R417

Am I missing sth?

Reply:
Yes.

What should be done in the seat: calling wl_keyboard_send_modifiers.

What should be done in wlr_keyboard: updating keyboard->modifiers.